specification/0057-configuration-metadata-api.md

```diff
@@ -0,0 +1,230 @@
+  Optional<T> widgetStateToValue(boolean widgetState, ValidationDiagnostics diagnostics);
+  boolean valueToWidgetState(T value);
+}
+
+MetadataType<ToggleValueConverter<?>, ?> TOGGLE_VALUE_CONVERTER;
+
+public interface CycleValueConverter<T> {
+  Optional<T> widgetStateToValue(int widgetState, ValidationDiagnostics diagnostics);
+  int valueToWidgetState(T value);
+  int valueCount();
+  int hiddenValueCount();
+}
+
+MetadataType<CycleValueConverter<?>, ?> CYCLE_VALUE_CONVERTER;
+
+// etc...
+```
+
+A Value Converter handles conversion between the stored value in a Setting and the canonical widget state. It also defines properties necessary for the widget to
+properly handle user input. For example, toggle widgets' state is a boolean value (checked or unchecked), while the stored values in the Setting may be any two
+values of any type.
+
+### Formatter
+
+```java
+@FunctionalInterface
+public interface Formatter<T> {
+  Text format(T value);
+}
+
+MetadataType<Formatter, ?> FORMATTER;
+```
+
+A Formatter is a function that converts a Setting's stored value into Text for displaying to players. This text is not meant to be mapped back to
+the value that produced it.
+
+#### Label
+
+```java
+MetadataType<Text, ?> LABEL;
+```
+
+A Setting's Label is the simple display name of a Setting.
+
+#### Tooltip
+
+```java
+@FunctionalInterface
+public interface TooltipSupplier<T> {
+  Text getTooltip(T value);
+}
+
+MetadataType<TooltipSupplier<?>, ?> TOOLTIP;
+```
+
+A Setting's tooltip is a brief description of the current state of the Setting, dependent on the Setting's value.
+
+#### Categories
+
+```java
+MetadataType<Collection<Identifier>, ?> CATEGORIES;
+```
+
+A Category is a logical grouping of Settings; for example, video, audio, accessibility, or gameplay. A Setting may be a member of any number
+of Categories. Categories are intended to be used to partition Settings into broad groupings for display purposes.
+
+This RFC defines a standard Category `c:accessibility`, to be used for Settings that are used to make a mod more accessible.
+
+#### Setting Validator
+
+Settings will use the existing `Constaint<T>` interface in Quilt Loader's configuration API to define Setting Validators.
+
+#### Group Validator
+
+```java
+@FunctionalInterface
+public interface GroupValidator {
+  void validate(ValidationDiagnostics diagnositics);
+  boolean equals(Object o);
+}
+
+MetadataType<GroupValidator, ?> GROUP_VALIDATOR;
+```
+
+A Group Validator is used to validate properties that pretain to the relationship among multiple Settings. For example, two Settings that
+represent the minimum and maximum value of a range may have a Group Validator that ensures the maximum is always greater than or equal to
+the minimum. It is intended that Group Validators that perform the same validation compare equal so that consumers of this API do not
+run validation logic multiple times.
+
+#### Environment Policy
+
+```java
+public enum EnvironmentPolicy {
```

Comment:
`CLIENT_ONLY` and `SERVER_ONLY` are functionally identical. I would therefore rename this class to `SyncPolicy` or something along those lines and condense those two options to `NONE`, with `SERVER_SYNCED` becoming `S2C` and `CLIENT_SYNCED` becoming `C2S`. Additionally, I might add a `P2P` sync type that shares user configs with other clients in a way that allows them to be accessed on a per-player basis. This has a lot of valuable implications for user-set display information (nicknames come to mind).